### ISSUE TITLE
Using Rails cache to cache results of graphql-batch promises?

### ISSUE THREAD

**Issue by clader** [Shopify/graphql-batch#73]
I currently have graphql-ruby acting as a front end for our LDAP directory.  graphql-batch works really great against directory data.  I've even gone so far as to cache the ldap query results that are formed by the batch loaders using Rails.cache statement.

But I'd like the caching to be a bit more granular (i.e. cache individual objects, and not entire sets of objects returned by batch queries.

The problem is when I try to wrap an ObjectLoader.for statement in Rails.cache.fetch block I get an 
```TypeError (no _dump_data is defined for class Proc):```
error.

This seems be because Rails.cache can't cache a promise.  Any thoughts or approaches on how to cache the results of an individual fulfill from a batch loader?

Heres some sample code from one of my resolvers that causes the above error:
```ruby 
      def call(obj, args, ctx)

		# calculate environment from the top level env ctx
		myEnv = get_root_query_env(ctx,args)

		# lets setup a cache key for this LDAP query
		myCacheKey = "ldap/#{myEnv}/#{args[:dn]}/#{Base64.encode64(LDAP::LDAP_USER_READABLE_ATTRS.to_s)}"
		Rails.logger.debug "Checking Rails Cache for key for LDAP User DN: #{args[:dn]}"

		# get cached result or if not cached, do actual query
		Rails.cache.fetch(myCacheKey, expires_in: LDAP::LDAP_GLBL_CACHE_EXPIRE_MINS.minutes) do

			# send the request for this user DN to the batch loader
			get_user_from_batch(myEnv, args[:dn])
		end
	end

	private

	def get_user_from_batch( myEnv, dn)
		# Call Generic directory Object loader so that this query can be batched
		Dir::Loaders::ObjectLoader.for("ldap", myEnv, "user").load(dn).then(nil, lambda do |exc|
				# looks like this query errored or wasn't found
				Rails.logger.error exc 
				GraphQL::ExecutionError.new("#{exc}: Object Not Found")
			end ) do |thisObj|
				# return the resolved object from the batch loader
				thisObj
		end
	end
```

**Comment by dylanahsmith** [comment]
> But I'd like the caching to be a bit more granular (i.e. cache individual objects, and not entire sets of objects returned by batch queries.

I think you still want to do the caching in the batch loader.  Instead of using a single Rails.cache.fetch block in the perform method, you could use `Rails.cache.read_multi`, batch load the uncached objects, then `Rails.cache.write` the loaded objects.  We have a [fetch_multi method in identity_cache](https://github.com/Shopify/identity_cache/blob/v0.5.1/lib/identity_cache/fallback_fetcher.rb#L21) that seems to be what you want, assuming you aren't manually deleting cache entries.

**Comment by clader** [comment]
I actually ended up doing both (i.e. I cache the results of a single batched query as a set of objects), but then I also added each individual object into the cache as the loader fulfilled it.  

So basically I just did away with the Rails.cache.fetch method and split the Rails.cache.write apart from the Rails.cache.exist/read.  So the Loader writes individual objects to the cache as they are fulfilled, and the Resolver that would send a query to the BatchLoader checks for the item in cache first.

This method seems to work quite well as the cache write/read are fully decoupled.
